editor/src/components/canvas/plugins/style-plugins.ts

```diff
@@ -1,33 +1,227 @@
+import { assertNever } from '../../../core/shared/utils'
+import * as EP from '../../../core/shared/element-path'
+import type { ValueAtPath } from '../../../core/shared/jsx-attributes'
+import type { EditorStateWithPatch } from '../commands/utils/property-utils'
+import { applyValuesAtPath } from '../commands/utils/property-utils'
+import * as PP from '../../../core/shared/property-path'
+import { emptyComments, jsExpressionValue } from '../../../core/shared/element-template'
+import { isStyleInfoKey, type StyleInfo } from '../canvas-types'
+
+export interface UpdateCSSProp {
+  type: 'set'
+  property: string
+  value: string | number
+}
+
+interface DeleteCSSProp {
+  type: 'delete'
+  property: string
+}
+
+export const updateCSSProp = (property: string, value: string | number): UpdateCSSProp => ({
+  type: 'set',
+  property: property,
+  value: value,
+})
+
+export const deleteCSSProp = (property: string): DeleteCSSProp => ({
+  type: 'delete',
+  property: property,
+})
+
+export type StyleUpdate = UpdateCSSProp | DeleteCSSProp
 
 export interface StylePlugin {
   name: string
   styleInfoFactory: StyleInfoFactory
-  normalizeFromInlineStyle: (
+  updateStyles: (
     editorState: EditorState,
-    elementsToNormalize: ElementPath[],
-    propertiesToRemove: PropertiesWithElementPath[],
-  ) => EditorState
+    elementPath: ElementPath,
+    updates: StyleUpdate[],
+  ) => EditorStateWithPatch
 }
 
-export const Plugins = {
-  InlineStyle: InlineStylePlugin,
-  Tailwind: TailwindPlugin,
-} as const
-
 export function getActivePlugin(editorState: EditorState): StylePlugin {
   if (isFeatureEnabled('Tailwind') && isTailwindEnabled()) {
     return TailwindPlugin(getTailwindConfigCached(editorState))
   }
   return InlineStylePlugin
 }
+
+export interface StylePropsUpdatedDuringInteraction {
+  propertiesUpdated: string[]
+  propertiesDeleted: string[]
+}
+
+export interface UpdatedProperties {
+  [elementPathString: string]: StylePropsUpdatedDuringInteraction
+}
+
+export function resetUpdatedProperties(editorState: EditorState): EditorState {
+  return { ...editorState, propertiesUpdatedDuringInteraction: {} }
+}
+
+function getPropertiesUpdatedDuringInteraction(editorState: EditorState) {
+  return editorState.propertiesUpdatedDuringInteraction
+}
+
+function ensureElementPathInUpdatedPropertiesGlobal(
+  elementPath: ElementPath,
+  updatedProperties: UpdatedProperties,
+) {
+  const updatedPropertiesToExtend = { ...updatedProperties }
+  const elementPathString = EP.toString(elementPath)
+  if (!(elementPathString in updatedPropertiesToExtend)) {
+    updatedPropertiesToExtend[elementPathString] = {
+      propertiesDeleted: [],
+      propertiesUpdated: [],
+    }
+  }
+
+  return updatedPropertiesToExtend
+}
+
+function runStyleUpdateMidInteraction(
+  editorState: EditorState,
+  elementPath: ElementPath,
+  updates: StyleUpdate[],
+): { editorStateWithChanges: EditorState; editorStatePatches: EditorStatePatch[] } {
```

Comment:
Is this the return type `EditorStateWithPatch`?

Reply:
`EditorStateWithPatch` only has a single patch. I factored out the return type into it's own interface, `EditorStateWithPatches`